webpack/assets/javascripts/jquery.ui.custom_spinners.js

```diff
@@ -0,0 +1,115 @@
+const megabyte = 1024 * 1024;
+const gigabyte = 1024 * megabyte;
+
+$(function () {
+  $.widget('ui.limited_spinner', $.ui.spinner, {
+    options: {
+      softMaximum: 0,
+      errorTarget: null
+    },
+    _validate: function () {
+      if (this.options.softMaximum !== 0) {
+        if (this.value() > this.options.softMaximum) {
+          this.options.errorTarget.show();
+        } else {
+          this.options.errorTarget.hide();
+        }
+      }
+    },
+    _spin: function (step, event) {
+      let result = this._super(step, event);
+
+      this._validate();
+      return result;
+    }
+  });
+
+  $.widget('ui.byte_spinner', $.ui.limited_spinner, {
+    options: {
+      step: 1,
+      min: 1,
+      incremental: false,
+      valueTarget: null
+    },
+    updateValueTarget: function () {
+      $('#' + this.options.valueTarget).val(this.value());
+    },
+    _gigabyteSpin: function (step) {
+      if (step > 0) {
+        if (step % gigabyte === 0) {
+          step = gigabyte;
+        } else {
+          step = gigabyte - (this.value() % gigabyte);
+        }
+      } else if (step < 0) {
+        if (this.value() % gigabyte === 0) {
+          step = -1 * gigabyte;
+        } else {
+          step = -1 * (this.value() % gigabyte);
+        }
+      }
+      return step;
+    },
+    _megabyteSpin: function (step) {
+      let megabyteStep = step * 256 * megabyte;
+
+      if (this.value() + megabyteStep > gigabyte) {
+        step = gigabyte - this.value();
+      } else if (this.value() + megabyteStep < megabyte) {
+        step = step * (this.value() - megabyte);
+      } else if (this.value() === megabyte && step > 0) {
+        step = 255 * megabyte;
+      } else {
+        step = megabyteStep;
+      }
+      return step;
+    },
+    _spin: function (step, event) {
+      let result = null;
+
+      if ((this.value() > gigabyte && step < 0) ||
+        (this.value() >= gigabyte && step > 0)) {
+        step = this._gigabyteSpin(step);
+      } else {
+        step = this._megabyteSpin(step);
+      }
+
+      result = this._super(step, event);
```

Comment:
It does not but `_spin` must be called before we update the hidden which holds the real value, so we need to call super and then the update. But I want to keep method interface therefore returning the result of super call.